### PR TITLE
Drawer implementation

### PR DESCRIPTION
**Pull request by lukelooker**
### :sparkles: Changes

- Drawer & useDrawer

### :white_check_mark: Requirements

- [ ] Includes test coverage for all changes
- [x] Build and tests are passing
- [x] Update documentation
- [x] Updated CHANGELOG
- [ ] Checked for i18n impacts
- [x] Checked for a11y impacts
- [x] Check for image-snapshot changes (run `yarn image-snapshots` locally)
- [ ] PR is ideally < ~400LOC

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -18,6 +18,7 @@
   "cSpell.words": [
     "JIRA",
     "camelcase",
+    "chuse",
```
Comment (mdodgelooker) [inline .vscode/settings.json:21]: 😍

File: package.json [hunk 2]
```diff
@@ -44,7 +44,7 @@
     "test": "yarn jest",
     "storybook": "yarn workspace storybook start",
     "storybooks-build": "yarn lerna run storybook-build --scope '@looker/*' --stream --parallel",
-    "preimage-snapshots": "yarn storybooks-build",
+    "-preimage-snapshots": "yarn storybooks-build",
```
Comment (mdodgelooker) [inline package.json:47]: To make it not always run before `image-snapshots`?

File: www/src/documentation/components/dialogs/drawer.mdx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+```jsx
+;() => {
+  const { drawer, open } = useDrawer({
+    content: 'Drawer Content',
+  })
+
+  return (
+    <>
+      {drawer}
+      <ButtonOutline onClick={openDialog}>useConfirm</ButtonOutline>
```
Comment (mdodgelooker) [inline www/src/documentation/components/dialogs/drawer.mdx:38]:
```suggestion
      <ButtonOutline onClick={open}>useDrawer</ButtonOutline>
```

File: packages/components/src/Drawer/Drawer.test.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    expect(screen.queryByText('Drawer content')).toBeInTheDocument()
+    const doneButton = screen.getByText('Done')
+    fireEvent.click(doneButton)
+    await waitForElementToBeRemoved(() => screen.getByText('Drawer content'))
+  })
+
+  test('DialogContext working', () => {
+    renderWithTheme(<Drawer defaultOpen content={<SimpleContent />} />)
+    expect(screen.queryByText('Drawer content')).toBeInTheDocument()
+  })
```
Comment (mdodgelooker) [inline packages/components/src/Drawer/Drawer.test.tsx:67]: Is this one still needed?

File: packages/components/src/Drawer/Drawer.tsx [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+export const useDrawer = ({
+  content,
+  defaultOpen = false,
+  width = '30rem',
+  canClose,
+  onClose,
+  ...props
+}: UseDrawerProps) => {
+  const [uncontrolledIsOpen, setControlledIsOpen] = useState(defaultOpen)
```
Comment (mdodgelooker) [inline packages/components/src/Drawer/Drawer.tsx:114]:
```suggestion
  const [uncontrolledIsOpen, setUncontrolledIsOpen] = useState(defaultOpen)
```

File: packages/components/src/Drawer/Drawer.tsx [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    callbackRef: focusRef,
+    disable: disableFocusTrap,
+    enable: enableFocusTrap,
+    isEnabled: focusTrapEnabled,
+    trapRef: focusTrapRef,
+  } = useFocusTrap(isOpen)
+  const [, portalRef] = useScrollLock(focusRef)
+
+  const handleOpen = () => {
+    !isControlled && setControlledIsOpen(true)
```
Comment (mdodgelooker) [inline packages/components/src/Drawer/Drawer.tsx:133]:
```suggestion
    !isControlled && setUncontrolledIsOpen(true)
```

File: packages/components/src/Drawer/Drawer.tsx [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  const handleClose = () => {
+    if (canClose && !canClose()) return
+    !isControlled && setControlledIsOpen(false)
```
Comment (mdodgelooker) [inline packages/components/src/Drawer/Drawer.tsx:138]:
```suggestion
    !isControlled && setUncontrolledIsOpen(false)
```

**Review by mdodgelooker — CHANGES_REQUESTED** [review]
Looks great! Just a naming change for the state setter, and a fix in the mdx. 👍

File: packages/components/src/Drawer/Drawer.test.tsx (hunk 4 above)
Comment (lukelooker) [reply packages/components/src/Drawer/Drawer.test.tsx:67]: Nope, that's silly

File: www/src/documentation/components/dialogs/drawer.mdx (hunk 3 above)
Comment (lukelooker) [reply www/src/documentation/components/dialogs/drawer.mdx]: Done

File: package.json (hunk 2 above)
Comment (lukelooker) [reply package.json]: Done

File: packages/components/src/Drawer/Drawer.tsx (hunk 6 above)
Comment (lukelooker) [reply packages/components/src/Drawer/Drawer.tsx]: Done

**Review by mdodgelooker — APPROVED** [review]
Ship it!

File: www/src/documentation/components/dialogs/drawer.mdx [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+```jsx
+;() => {
+  const { drawer, open } = useDrawer({
+    content: 'Drawer Content',
+  })
+
+  return (
+    <>
+      {drawer}
+      <ButtonOutline onClick={open}>useConfirm</ButtonOutline>
```
Comment (mdodgelooker) [inline www/src/documentation/components/dialogs/drawer.mdx:32]:
```suggestion
      <ButtonOutline onClick={open}>useDrawer</ButtonOutline>
```

**Review by mdodgelooker — CHANGES_REQUESTED** [review]

**Review by mdodgelooker — APPROVED** [review]

**Pull request merged**